### ISSUE TITLE
"invalid next size" at start

### ISSUE THREAD

**Issue by asfluido** [mruby/mruby#2758]
Hello Mruby developers.

I am currently developing a quite consistent application in Mruby, under Linux. The application makes extensive use of C-Ruby interaction, and has both C threads and multiple virtual machines. I am very happy about how the development is proceeding.

I currently make use of the Mruby code checked out of github in the morning of 27 March. 

I am encountering a couple of sporadic problems. I will open an issue for each, so discussion can proceed in parallel.

This first problem happens when I bring up my application. It happens once every 5-10 runs, and tends to happen in short bursts (twice or thrice in sequence). It started to appear when the complexity increased, but I could not track it to some specific block of code being added. The crash message is as follows:

> *** Error in `mruby/bin/mruby': free(): invalid next size (fast): 0x0000000002828ec0 ***

Here is the process stack obtained from the core:

```
#0  0x00007fb8a9a9bfd7 in raise () from /lib/libc.so.6
#1  0x00007fb8a9a9d3b8 in abort () from /lib/libc.so.6
#2  0x00007fb8a9ad9f64 in ?? () from /lib/libc.so.6
#3  0x00007fb8a9adf73e in ?? () from /lib/libc.so.6
#4  0x00007fb8a9ae0446 in ?? () from /lib/libc.so.6
#5  0x0000000000419987 in mrb_default_allocf (mrb=0xd38010, p=0xeb3110, size=0, ud=0x0) at /data/code/motri/mruby/src/state.c:60
#6  0x0000000000410376 in mrb_free (mrb=0xd38010, p=0xeb3110) at /data/code/motri/mruby/src/gc.c:244
#7  0x0000000000411037 in obj_free (mrb=0xd38010, obj=0xe76860) at /data/code/motri/mruby/src/gc.c:652
#8  0x0000000000411797 in incremental_sweep_phase (mrb=0xd38010, limit=18446744073709551615) at /data/code/motri/mruby/src/gc.c:874
#9  0x0000000000411a1a in incremental_gc (mrb=0xd38010, limit=18446744073709551615) at /data/code/motri/mruby/src/gc.c:935
#10 0x0000000000411a76 in incremental_gc_until (mrb=0xd38010, to_state=GC_STATE_ROOT) at /data/code/motri/mruby/src/gc.c:951
#11 0x0000000000411c67 in mrb_incremental_gc (mrb=0xd38010) at /data/code/motri/mruby/src/gc.c:1000
#12 0x0000000000410834 in mrb_obj_alloc (mrb=0xd38010, ttype=MRB_TT_STRING, cls=0xd42600) at /data/code/motri/mruby/src/gc.c:405
#13 0x0000000000420d99 in str_new (mrb=0xd38010, p=0x0, len=0) at /data/code/motri/mruby/src/string.c:168
#14 0x0000000000422bbf in mrb_str_dup (mrb=0xd38010, str=...) at /data/code/motri/mruby/src/string.c:740
#15 0x000000000044649b in mrb_context_run (mrb=0xd38010, proc=0xe1e5a0, self=..., stack_keep=0) at /data/code/motri/mruby/src/vm.c:2155
#16 0x0000000000447452 in mrb_toplevel_run_keep (mrb=0xd38010, proc=0xe1d940, stack_keep=0) at /data/code/motri/mruby/src/vm.c:2400
#17 0x000000000045685b in load_exec (mrb=0xd38010, p=0xe3ccd0, c=0xe3cab0) at src/parse.y:5599
#18 0x00000000004568d1 in mrb_load_file_cxt (mrb=0xd38010, f=0xe3c830, c=0xe3cab0) at src/parse.y:5608
#19 0x000000000040fc89 in main (argc=3, argv=0x7fffa7e7f648) at /data/code/motri/mruby/mrbgems/mruby-bin-mruby/tools/mruby/mruby.c:227
```

The crash happens when trying to free an object while allocating other stuff. I noticed that the object is an Array (see gc.c 652) but its pointer (`((struct RArray*)obj)->ptr`) is NULL. 

I tried to patch the behaviour by changing the body of the `mrb_free` function from

```
(mrb->allocf)(mrb, p, 0, mrb->allocf_ud);
```

to

```
 if(p)
   (mrb->allocf)(mrb, p, 0, mrb->allocf_ud);
```

and I found out that I still had crashes happening, when the pointer was not NULL, but the length (`mrb->allocf_ud`) was 0. So I further patched `mrb_free` as follows:

```
 if(p && mrb->allocf_ud>0)
   (mrb->allocf)(mrb, p, 0, mrb->allocf_ud);
```

I had no further crashes, but the memory usage of the process quickly reached 40-50 %, until I had to kill the process (when the process starts correctly, it won't occupy more than 8-9% (of 12GB)).  So I reverted to the original code.

I am clueless about what to do next.  I will gladly try whatever you suggest, and report back.

Carlo

**Comment by take-cheeze** [comment]
About
> length (mrb->allocf_ud) 

The 3rd argument of `mrb->allocf` call is the memory size and the 4th argument is a user data for allocator which isn't used in mruby's default allocator.
So checking `mrb->allocf_ud` is irrelevant unless you use it.

BTW, mruby allocator (`mrb->allocf`)
* frees given pointer if size is 0
* `malloc`s given size if given pointer is `nullptr`

**Comment by cremno** [comment]
Valgrind, or compile your project with a recent GCC or Clang version and the [AddressSanitizer](http://clang.llvm.org/docs/AddressSanitizer.html).

By the way, passing a null pointer to `free()` and even `mrb_free()` is totally fine since C89.

**Comment by asfluido** [comment]
I went the way of AddressSanitizer. What I did:

* I set CFLAGS to `-O1 -g -march=native -fsanitize=address -fno-omit-frame-pointer` (usually set to `-O3 -g -march=native`). 
* I set LDFLAGS to `-fsanitize=address -fno-omit-frame-pointer` (usually unset)
* I ran `make clean`, then removed the `build` directory, then ran make. 

I ran the program a dozen times. The error that I mention in this thread has never happened, but every time I kill the process (with Ctrl-C) I receive a message from AddressSanitizer:

```
=================================================================
==30882== ERROR: AddressSanitizer: heap-buffer-overflow on address 0x6066000160d0 at pc 0x48e5fe bp 0x7ffc39bd2b50 sp 0x7ffc39bd2b48
READ of size 4 at 0x6066000160d0 thread T0
    #0 0x48e5fd (/data/code/motri/mruby/bin/mruby+0x48e5fd)
    #1 0x49ccb7 (/data/code/motri/mruby/bin/mruby+0x49ccb7)
    #2 0x4c47e4 (/data/code/motri/mruby/bin/mruby+0x4c47e4)
    #3 0x4c48a3 (/data/code/motri/mruby/bin/mruby+0x4c48a3)
    #4 0x411162 (/data/code/motri/mruby/bin/mruby+0x411162)
    #5 0x7f8faff7dad4 (/lib/libc-2.19.so+0x21ad4)
    #6 0x40fad3 (/data/code/motri/mruby/bin/mruby+0x40fad3)
0x6066000160d0 is located 48 bytes to the left of 4960-byte region [0x606600016100,0x606600017460)
allocated by thread T0 here:
    #0 0x7f8fb45eb54f (/usr/lib/libasan.so.0.0.0+0x1554f)
    #1 0x42b493 (/data/code/motri/mruby/bin/mruby+0x42b493)
    #2 0x41210b (/data/code/motri/mruby/bin/mruby+0x41210b)
    #3 0x41220e (/data/code/motri/mruby/bin/mruby+0x41220e)
    #4 0x4868bf (/data/code/motri/mruby/bin/mruby+0x4868bf)
    #5 0x48ec1f (/data/code/motri/mruby/bin/mruby+0x48ec1f)
    #6 0x49cb70 (/data/code/motri/mruby/bin/mruby+0x49cb70)
    #7 0x488500 (/data/code/motri/mruby/bin/mruby+0x488500)
    #8 0x4360a8 (/data/code/motri/mruby/bin/mruby+0x4360a8)
    #9 0x48f056 (/data/code/motri/mruby/bin/mruby+0x48f056)
    #10 0x49ccb7 (/data/code/motri/mruby/bin/mruby+0x49ccb7)
    #11 0x4c47e4 (/data/code/motri/mruby/bin/mruby+0x4c47e4)
    #12 0x4c48a3 (/data/code/motri/mruby/bin/mruby+0x4c48a3)
    #13 0x411162 (/data/code/motri/mruby/bin/mruby+0x411162)
    #14 0x7f8faff7dad4 (/lib/libc-2.19.so+0x21ad4)
Shadow bytes around the buggy address:
  0x0c0d3fffabc0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0d3fffabd0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0d3fffabe0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0d3fffabf0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0d3fffac00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c0d3fffac10: fa fa fa fa fa fa fa fa fa fa[fa]fa fa fa fa fa
  0x0c0d3fffac20: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0d3fffac30: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0d3fffac40: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0d3fffac50: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c0d3fffac60: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte reprepsents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:     fa
  Heap righ redzone:     fb
  Freed Heap region:     fd
  Stack left redzone:    f1
  Stack mid redzone:     f2
  Stack right redzone:   f3
  Stack partial redzone: f4
  Stack after return:    f5
  Stack use after scope: f8
  Global redzone:        f9
  Global init order:     f6
  Poisoned by user:      f7
  ASan internal:         fe
==30882== ABORTING
```

Via GDB I extracted the locations of the frames:

\#0: `mruby/src/vm.c`, line 1033 (in `mrb_context_run`)
\#1: `mruby/src/vm.c`, line 2400 (in `mrb_toplevel_run_keep`)
\#2: `mruby/src/parse,y`, line 5599 (in `load_exec`)
\#3: `mruby/src/parse,y`, line 5608 (in `mrb_load_file_cxt`)
\#4: `mruby/mrbgems/mruby-bin-mruby/tools/mruby/mruby.c`, line 227 (in `main`)

The line that causes the error looks as follows:

```
      for (n=0; n<a && eidx > ci[-1].eidx; n++) {
```

I can't say much more. With sanitized loaded, the app runs more or less as fluid as without, and the crash at start never takes place.

Carlo

**Comment by asfluido** [comment]
I follow up on my previous comment because, in a certainly non-orthodox way, I was apparently able to solve both issues that I had posted about. 

AddressSanitizer was constantly complaining about heap buffer overflow, pointing its finger to line 1033 of vm.c.

The block reads as follows, line 1033 is marked with an arrow:

```
   CASE(OP_EPOP) {
      /* A      A.times{ensure_pop().call} */
      int a = GETARG_A(i);
      mrb_callinfo *ci = mrb->c->ci;
      int n, eidx = ci->eidx;
      
      for (n=0; n<a && eidx > ci[-1].eidx; n++) { /* <--- */
        ecall(mrb, --eidx);
        ARENA_RESTORE(mrb, ai);
      }
      NEXT;
    }
```

The complaint was always about accessing a location to the **left** of the assigned memory region. I know *very little* about how the Mruby virtual machine works, but that `[-1]` looked very unhealthy to me. I said, what the heck, the machine won't explode, and I substituted 
```
      for (n=0; n<a && eidx > ci[-1].eidx; n++) {
```
with
```
      for (n=0; n<a && eidx > ci[0].eidx; n++) {
```

Well, to my astonishment  all my crashes disappeared. The sporadic problems without AddressSanitizer that I described in my two issues also seem to have disappeared (I tried a largish handful of times).  My code runs smoothly, memory usage remains low, and the sun shines out of the window ;-)

Of course, I do not have the faintest idea about which obscure consequences this patch of mine may have. But I certainly will keep it in my local version, in the hope to soon find on these pages either a corroboration of my purely experimental results or a better fix. I can also submit a patch if given green light.
